### PR TITLE
Implement cloud functions. Implement OH stats.

### PR DESCRIPTION
**Pull request by abcdefguan**
### Summary 

Added cloud functions to track statistics

Todo:
Automatically add TAs to session upon updating a question state
Restore undo functionality

Done:
Increment number of answered questions when question becomes "resolved"
Increment number of total questions when question is added
Increment total response time when question becomes "resolved"
Increment total wait time when question becomes "assigned"
Increment number of waited for question when question becomes "assigned"

Changes:
Added Cloud functions to project
Deployed cloud function that increments totalQuestions when new question is asked => Tested to be working

Added a number of fields to session
(totalQuestions, assignedQuestions, resolvedQuestions, totalWaitTime, totalResolveTime)
totalQuestions => Total number of questions in session (excluding no-show and retracted)
assignedQuestions => Number of questions assigned to TAs in session
resolvedQuestions => Number of questions resolved by TAs in session
totalWaitTime => Summation of waiting times for all questions (Time between question asked and assignment) in seconds
totalResolveTime => Summation of resolution times for all questions (Time between question assigned and resolved) in seconds
Added timeAssigned field in questions for tracking response time
Made timeAssigned and timeAddressed "public" => Reverted because this is not needed for the cloud function approach
Added default values for all of these fields to session.

The following variables are now automatically tracked with a cloud function and will update based on state changes:
totalQuestions
assignedQuestions
resolvedQuestions
totalWaitTime
totalResolveTime

Implemented rudimentary frontend interface to display OH stats.

Typescript is working correctly in functions folder

### Test Plan 

Checked that the database values of the variables above were correct when questions were assigned and removed

### Breaking Changes 

None

Firebase schema change (requires migration plan)
=> No migration necessary. Firebase increment automatically generates field.
I updated existing types in `/src/components/types/`
My changes requires a change to the documentation.
Other change that could cause problems (Detailed in notes)
Changes to Cloud Functions

### Checklist

- [ ] I have updated the documentation accordingly.
- [ ] My PR adds a @ts-ignore

### CODE REVIEW

File: functions/tslint.json [hunk 1]
```diff
@@ -0,0 +1,115 @@
+{
```
Comment (SamChou19815) [inline functions/tslint.json:1]: Don't need to include tslint. We already migrated to eslint with TS support

File: functions/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        return db.doc(`sessions/${sessionId}`).update({
+            totalQuestions: increment
+        });
+    });
+
+// This map maps a question status to a tuple of
+// 1. The number of questions this counts as (0 if no show or retracted)
+// 2. The number of assigned questions this counts as (1 if assigned / resolved)
+// 3. The number of resolved questions this counts as (1 if resolved only)
+const questionStatusNumbers = new Map<string, number[]>();
```
Comment (SamChou19815) [inline functions/src/index.ts:40]: Use tuple type `[number, number, number]` instead

File: functions/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
+import * as functions from 'firebase-functions';
+import * as admin from 'firebase-admin';
```
Comment (SamChou19815) [inline functions/src/index.ts:2]: To fix the linter error here, add a yarn install step in CI for the functions folder

File: functions/package.json [hunk 4]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (SamChou19815) [inline functions/package.json:1]: Why there is no yarn.lock for this package.json file

**Review by SamChou19815 — DISMISSED** [review]

File: functions/src/index.ts (hunk 3 above)
Comment (abcdefguan) [reply functions/src/index.ts:2]: I don't know how to do this. The .github/workflows file is read from the root directory only. Also, yarn cannot specify a package.json from some other directory. Could you please help with this?

**Comment by abcdefguan** [comment]
@SamChou19815 
functions/index.ts doesn't use Typescript because I had trouble importing fireData.d.ts. Do you know how to do this correctly?

**Comment by abcdefguan** [comment]
Sorry, pressed wrong button

File: firebase.json [hunk 5]
```diff
@@ -1,12 +1,22 @@
...
         "source": "**",
         "destination": "/index.html"
       }
     ]
+  },
+  "functions": {
+    "predeploy": [
+      "npm --prefix \"$RESOURCE_DIR\" run lint",
+      "npm --prefix \"$RESOURCE_DIR\" run build"
+    ]
```
Comment (SamChou19815) [inline firebase.json:20]: Replace that with equivalent yarn command.

File: functions/src/index.ts (hunk 3 above)
Comment (SamChou19815) [reply functions/src/index.ts:2]: CI config fixed.

**Comment by SamChou19815** [comment]
> functions/index.ts doesn't use Typescript because I had trouble importing fireData.d.ts. Do you know how to do this correctly?

This will be more involved. The most correct way is to use yarn workspace to allow cross-package code sharing, but the setup is tricky since firebase doesn't support yarn workspace well. Meanwhile you have two options:

- Keep it as it is
- Copy over type definitions that you will use.

**Review by ngwattcos — APPROVED** [review]
Thank you for these changes. This is a substantial update for Queue Me In that allows us to implement important features like improved TA promotion and performance tracking, among other features that rely on Cloud Functions. Nice work on the CI/CD.

**Review by ngwattcos — COMMENTED** [review]
Looks great!

**Pull request merged**